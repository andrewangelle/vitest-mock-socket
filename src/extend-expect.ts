import { expect } from 'vitest';

import { toHaveReceivedMessages } from './matchers/toHaveReceivedMessages';
import { toHaveResolvedMessages } from './matchers/toHaveResolvedMessages';
import { toReceiveMessage } from './matchers/toReceiveMessage';
import type { DeserializedMessage, MessageMatcherOptions } from './types';

interface CustomMatchers<MatcherResult = unknown> {
  toReceiveMessage<MessageType = object>(
    message: DeserializedMessage<MessageType>,
    options?: MessageMatcherOptions,
  ): Promise<MatcherResult>;
  toHaveReceivedMessages<MessageType = object>(
    messages: DeserializedMessage<MessageType>[],
  ): MatcherResult;
  toHaveResolvedMessages<MessageType = object>(
    messages: DeserializedMessage<MessageType>[],
    options?: MessageMatcherOptions,
  ): Promise<MatcherResult>;
}

declare module '@vitest/expect' {
  // biome-ignore lint/suspicious/noExplicitAny: required to extend vitest matchers
  interface Assertion<T = any> extends CustomMatchers<T> {}
  interface AsymmetricMatchersContaining extends CustomMatchers {}
}

expect.extend({
  toHaveReceivedMessages,
  toReceiveMessage,
  toHaveResolvedMessages,
});
