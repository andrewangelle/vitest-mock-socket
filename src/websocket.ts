import type { Client, CloseOptions } from 'mock-socket';
import { Server } from 'mock-socket';

import type { DeserializedMessage, WebSocketServerOptions } from './types';

export class WebSocketServer {
  static clients: WebSocketServer[] = [];

  static clean() {
    for (const client of WebSocketServer.clients) {
      client.close();
      client.messages = [];
    }
    WebSocketServer.clients = [];
  }

  server: Server;
  messages: DeserializedMessage[] = [];
  #options: WebSocketServerOptions;
  #isConnected = Promise.withResolvers<Client>();
  #isClosed = Promise.withResolvers<Client>();
  #messageAdded = Promise.withResolvers<void>();
  #messageQueue: DeserializedMessage[] = [];

  constructor(
    url: string,
    options: WebSocketServerOptions = { jsonProtocol: false },
  ) {
    WebSocketServer.clients.push(this);
    this.#options = options;
    this.server = this.#createServer(url);
  }

  async connected() {
    return this.#isConnected.promise;
  }

  async closed() {
    return this.#isClosed.promise;
  }

  async nextMessage() {
    // if we have a queued message return it immediately
    if (this.#messageQueue.length > 0) {
      return this.#messageQueue.shift();
    }

    // otherwise wait until we have one
    await this.#messageAdded.promise;
    return this.#messageQueue.shift();
  }

  on(
    eventName: 'connection' | 'message' | 'close',
    callback: (socket: Client) => void,
  ): void {
    this.server.on(eventName, callback);
  }

  send(message: DeserializedMessage) {
    this.server.emit('message', this.#serializer(message));
  }

  close(options?: CloseOptions) {
    this.server.close(options);
  }

  error(options?: CloseOptions) {
    this.server.emit('error', null);
    this.server.close(options);
  }

  #serializer(deserializedMessage: DeserializedMessage) {
    if (this.#options.jsonProtocol) {
      return JSON.stringify(deserializedMessage);
    }
    return deserializedMessage;
  }

  #deserializer(serializedMessage: string): DeserializedMessage {
    if (this.#options.jsonProtocol) {
      try {
        const safeParsed = JSON.parse(serializedMessage);
        return safeParsed;
      } catch {
        // silently swallow JSON parse error
      }
    }
    return serializedMessage;
  }

  #createServer(url: string) {
    const server = new Server(url, this.#options);

    server.on('close', (socket: Client) => {
      this.#isClosed.resolve(socket);
      this.#isClosed = Promise.withResolvers();
    });

    server.on('connection', (socket: Client) => {
      this.#isConnected.resolve(socket);
      this.#isConnected = Promise.withResolvers();

      socket.on('message', (message) => {
        const parsedMessage = this.#deserializer(message as string);
        this.messages.push(parsedMessage);
        this.#messageQueue.push(parsedMessage);
        this.#messageAdded.resolve();
        this.#messageAdded = Promise.withResolvers();
      });
    });

    return server;
  }
}
