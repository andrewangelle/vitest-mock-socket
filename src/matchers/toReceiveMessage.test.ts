import '../extend-expect';

import { WebSocketServer } from '../websocket';

let server: WebSocketServer;
let client: WebSocket;

const testURL = 'ws://localhost:1234';
beforeEach(async () => {
  server = new WebSocketServer(testURL);
  client = new WebSocket(testURL);
  await server.connected();
});

afterEach(() => {
  WebSocketServer.clean();
});

describe('.toReceiveMessage', () => {
  it('passes when the websocket server receives the expected message', async () => {
    client.send('hello there');
    await expect(server).toReceiveMessage('hello there');
  });

  it('passes when the websocket server receives the expected message with custom timeout', async () => {
    setTimeout(() => {
      client.send('hello there');
    }, 2000);

    await expect(server).toReceiveMessage('hello there', { timeout: 3000 });
  });

  it('passes when the websocket server receives the expected JSON message', async () => {
    client.send(`{"answer":42}`);
    await expect(server).toReceiveMessage({ answer: 42 });
  });

  it('passes with JSON protocol on, but message is not JSON', async () => {
    client.send('not json');
    await expect(server).toReceiveMessage('not json');
  });

  it('fails when called with an expected argument that is not a valid WS', async () => {
    expect.hasAssertions();
    await expect(
      expect('boom').toReceiveMessage('hello there'),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('fails when the WS server does not receive the expected message', async () => {
    expect.hasAssertions();
    await expect(
      expect(server).toReceiveMessage('hello there'),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('fails when the WS server does not receive the expected message with custom timeout', async () => {
    expect.hasAssertions();
    await expect(
      expect(server).toReceiveMessage('hello there', { timeout: 3000 }),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('fails when the WS server receives a different message', async () => {
    expect.hasAssertions();
    client.send('hello there');
    await expect(
      expect(server).toReceiveMessage('HI!'),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('passes when expecting a JSON message but the server is not configured for JSON protocols', async () => {
    expect.hasAssertions();
    client.send(`{"answer":42}`);
    expect(server).toReceiveMessage({ answer: 42 });
  });
});

describe('.not.toReceiveMessage', () => {
  it("passes when the websocket server doesn't receive the expected message", async () => {
    client.send('hello there');
    await expect(server).not.toReceiveMessage("What's up?");
  });

  it('fails when called with an expected argument that is not a valid WS', async () => {
    expect.hasAssertions();
    await expect(
      expect('boom').not.toReceiveMessage('hello there'),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it("fails when the WS server doesn't receive any messages", async () => {
    expect.hasAssertions();
    await expect(
      expect(server).not.toReceiveMessage('hello there'),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('fails when the WS server receives the un-expected message', async () => {
    expect.hasAssertions();
    client.send('hello there');
    await expect(
      expect(server).not.toReceiveMessage('hello there'),
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
