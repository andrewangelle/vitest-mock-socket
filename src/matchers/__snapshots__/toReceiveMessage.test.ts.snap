// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`.not.toReceiveMessage > fails when called with an expected argument that is not a valid WS 1`] = `
[Error: [2mexpect([22m[31mWebSocketServer[39m[2m).not.toReceiveMessage([22m[32mexpected[39m[2m)[22m

Expected the websocket object to be a valid WebSocketServer mock.
Received: string
  [31m"boom"[39m]
`;

exports[`.not.toReceiveMessage > fails when the WS server doesn't receive any messages 1`] = `
[Error: [2mexpect([22m[31mWebSocketServer[39m[2m).not.toReceiveMessage([22m[32mexpected[39m[2m)[22m

Expected the websocket server to receive a message,
but it didn't receive anything in 1000ms.]
`;

exports[`.not.toReceiveMessage > fails when the WS server receives the un-expected message 1`] = `
[Error: [2mexpect([22m[31mWebSocketServer[39m[2m).not.toReceiveMessage([22m[32mexpected[39m[2m)[22m

Expected the next received message to not equal:

  [32m"hello there"[39m

Received:

  [31m"hello there"[39m
]
`;

exports[`.toReceiveMessage > fails when called with an expected argument that is not a valid WS 1`] = `
[Error: [2mexpect([22m[31mWebSocketServer[39m[2m).toReceiveMessage([22m[32mexpected[39m[2m)[22m

Expected the websocket object to be a valid WebSocketServer mock.
Received: string
  [31m"boom"[39m]
`;

exports[`.toReceiveMessage > fails when the WS server does not receive the expected message 1`] = `
[Error: [2mexpect([22m[31mWebSocketServer[39m[2m).toReceiveMessage([22m[32mexpected[39m[2m)[22m

Expected the websocket server to receive a message,
but it didn't receive anything in 1000ms.]
`;

exports[`.toReceiveMessage > fails when the WS server does not receive the expected message with custom timeout 1`] = `
[Error: [2mexpect([22m[31mWebSocketServer[39m[2m).toReceiveMessage([22m[32mexpected[39m[2m)[22m

Expected the websocket server to receive a message,
but it didn't receive anything in 3000ms.]
`;

exports[`.toReceiveMessage > fails when the WS server receives a different message 1`] = `
[Error: [2mexpect([22m[31mWebSocketServer[39m[2m).toReceiveMessage([22m[32mexpected[39m[2m)[22m

Expected the next received message to equal:

  [32m"HI!"[39m

Received:

  [31m"hello there"[39m

Difference:

  [32m- Expected[39m
  [31m+ Received[39m
  [32m- HI![39m
  [31m+ hello there[39m
]
`;
